include/envoy/network/connection.h

```diff
@@ -205,10 +205,50 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
    * @return boolean telling if the connection is currently above the high watermark.
    */
   virtual bool aboveHighWatermark() const PURE;
+
+  enum class PostIoAction { Close, KeepOpen };
+
+  struct IoResult {
+    PostIoAction action_;
+    uint64_t bytes_processed_;
+  };
+};
+
+class TransportSecurityCallbacks {
+public:
+  virtual ~TransportSecurityCallbacks(){};
+
+  virtual const Connection& connection() const PURE;
+
+  virtual void raiseEvent(ConnectionEvent event) PURE;
+
+  virtual bool shouldDrainReadBuffer() PURE;
+
+  virtual void setReadBufferReady() PURE;
+
+  virtual int fd() PURE;
+
+  virtual Buffer::Instance& readBuffer() PURE;
+
+  virtual Buffer::Instance& writeBuffer() PURE;
 };
 
 typedef std::unique_ptr<Connection> ConnectionPtr;
 
+class TransportSecurity {
+public:
+  virtual ~TransportSecurity() {}
+  virtual std::string nextProtocol() const PURE;
+  virtual Connection::IoResult doReadFromSocket() PURE;
```

Comment:
doReadAndDecrypt?

Reading from the socket is something I feel the connection should do.  Reading and transforming is more of a crypto operation

Reply:
Also I'm really not thrilled about removing the read from the connectionImpl and having the transport-security do it using the raw fd.  It makes it conceptually much harder to do something like https://github.com/envoyproxy/envoy/issues/1630 where you may have a stream reading from another stream rather than from a socket.  If the connection is supposed to represent the underlying byte stream it makes more conceptual sense to me to have the connection do the read and then pass it to the crypto layer which does decryption.  That unfortunately doesn't play at all well with how we're currently doing SSL_reads.  Is there a way to keep the logical "get data" separate from the decryption operation (without incurring extra copies)?

Reply:
I totally agree that not let transport-security do it using raw fd, and I would love to see that read/write from socket and encrypt/decrypt are decoupbled. But current SSL_do_handshake and SSL_read/write isn't in that way. I did this prototype is based on current implementation to pull out the interface.

I don't understand the part that #1630 will be harder, if you read from stream then I assume you'll use `Http::AsyncClient`, no?

@PiotrSikora do you have any idea?

Reply:
As written, I don't think this has anything to do with transport security. It's basically a pluggable connection handling module. This is potentially useful for people that want to port Envoy to proprietary environments. (This is basically my comment from the API PR).

If we have to go this direction, what do we think about just calling it what it is? Something like `LowLevelSocket` or whatever. Then we can have `PlainTextLowLevelSocket`, `BoringSslLowLevelSocket`, etc.?

Reply:
re: BoringSSL without fd, looking at gRPC TSI impl, it is doable with two `BIO_s_mem`, like what gRPC does [here](https://github.com/grpc/grpc/blob/master/src/core/tsi/ssl_transport_security.cc#L1104), 
then interface will be just shuffling buffers with methods:
- doHandshake
- encryptBuffer
- decryptBuffer

This needs more work on current BoringSSL implementation, in that way the interface is pure transport security. WDYT?

Reply:
I don't really have a problem with the interface you have created (at a super high level, haven't fully reviewed) if we treat it for what it currently is: a method to plug in a totally new connection type. If we want to just hit transport security we should do what you just suggested. I could go either way.

Reply:
Matt's comments are spot on.  I don't like "security interface" reading from a socket but if we rebrand it as socket helper or some such my objections go away :-)

If we go that route I think I'd have some other renaming in mind.  For example, I'd suggest renaming nextProtocol() at it implies NPN, but a simple protocol() would be more neutral (and also useful for the proxying-random-things-over-H2 use case I have in mind :-) )

Reply:
Yeah I agree that current shape is more like low level socket. I'll try a bit to narrow this down to a real "security interface" if it is easy. If not I would rename this and config to `LowLevelSocket` per suggested.

Reply:
@lizan FWIW, I think we want the low level socket change no matter what. We have two new connection implementations coming (Windows and fd.io/DPDK/VPP) which I think could make good use of this. Not sure if this sways your opinion or not.

Reply:
@mattklein123 After digging with BoringSSL a while, and looked winsock, etc, I reached same conclusion here with `LowLevelSocket`. So yes, I will update this with renaming / comments etc.

Reply:
I agree with @mattklein123  here.  We've got some exciting TCP acceleration coming down the pike from fd.io/VPP, and a pluggable LowLevelSocket would greatly ease that process :)